docs: expand contribution doc into a comprehensive section. (#4271)
wix/Detox: docs/contributing/code/setting-up-the-dev-environment.md

:::

## Setting Up The Monorepo Management

Our repository is a monorepo, which means it contains multiple Detox-related projects and packages. [Read more about our repository structure](../code/overview.md#repository-structure).

To set up the monorepo locally, follow these steps:

Install the monorepo management tool, `lerna`:

```bash npm2yarn
npm install lerna@3.x.x --global
```

Clone the repository and navigate to the project directory:

```bash
git clone git@github.com:wix/Detox.git
cd detox
git submodule update --init --recursive
```

From the project's root directory, install and link the internal projects:

```bash
lerna bootstrap
```

## Installing Common Dependencies

### React-Native CLI

[react-native-cli] is a command line interface for React Native.

```bash npm2yarn
npm install react-native-cli --global
```

### Watchman

[Watchman] is a tool by Facebook for watching changes in the filesystem.

```bash
brew install watchman
```

### xcpretty

[xcpretty] is a fast and flexible formatter for xcodebuild.

```bash
gem install xcpretty
```

> ## Building and Testing

Reviewer: This has to be split away onto a dedicate guide, as it's not about setting up the environment.

:::

The following sections describe how to build and test our code, with a general instruction.
However, we recommend you to refer to our `package.json` files for a complete list of available scripts:

- Detox Framework: [`detox/package.json`]
- Detox Test App: [`detox/test/package.json`]

### Unit Tests

We are using [Jest] for running our unit tests.

Under the `detox` directory, run the following command to run the unit tests with coverage:

```bash
npm run test
```

We also have unit tests for our native code (Android only), you can run them using the following commands:

```bash
npm run test:android-debug
npm run test:android-release
```

### iOS: Rebuilding the Framework

After changing the native code of Detox iOS, you need to rebuild the Detox framework. This is done when running:

```bash
detox build-framework-cache
```

Note that it is only required when you change the native code of Detox, or one of its dependencies (e.g. [DetoxSync]).
If you are only changing the JavaScript code, you don't need to rebuild the framework.

### Building the Detox Test App

The Detox test app is a full-feature React Native app for testing end-to-end testing.

> To build the Detox test app (`detox/test`), use the `build` command of the Detox CLI.

Reviewer: I'm having some thoughts here. Why not advise using npm scripts, instead of the detox cli directly? To some extent, npm scripts are abstractions over the actual command, and that might help make our docs more future proof. @noomorph wdyt?

List of available configurations can be found under [`detox/test/e2e/detox.config.js`].

For example, to build the Detox test app for iOS in release mode, run:

```bash
detox build --configuration ios.sim.release
```

### End-to-End Tests

To run the end-to-end tests, use the `test` command of the Detox CLI.

For example, to run the end-to-end tests for iOS in release mode, run:

> ```bash
> detox test --configuration ios.sim.release
> ```

Reviewer: ```suggestion
```bash
cd detox/test
detox test --configuration ios.sim.release
```

To run in Android release mode, run:
```bash
cd detox/test
detox test --configuration android.emu.release
```
> Mind that on Android, Detox expects a specific AVD with a [specific name](https://github.com/wix/Detox/blob/24f68c3bd7675ed6f508e3f5fea97509eb7991b2/detox/test/e2e/detox.config.js#L111) to be available on your computer. Be sure to [install such an AVD](https://developer.android.com/studio/run/managing-avds#createavd), beforehand.

### Integration Tests

> We are using [Jest] for running our integration tests.

Reviewer: ```suggestion
Besides unit tests and end-to-end tests, we have some middle ground integration tests, which typically run Detox in a custom headless-like mode (i.e. stubs replace devices). Those tests typically execute Detox from a command line in that mode, and then do some post-processing of the results. We use [Jest] for running our integration tests.

The integration reside under a dedicated directory alongside Detox's the self-test app's code.
```

Under the `detox/test` directory, run the following command to run the integration tests:

```bash
npm run integration
```

### Linting

> We are using [ESLint] for linting our JavaScript code.

Reviewer: ```suggestion
We use [ESLint] for linting our JavaScript code.
```

You can run the linter locally using the following command (under `detox/test`):

```bash
npm run lint
```

We also test our types using the following command:

```bash
npm run test:types
```

### Running the CI Scripts Locally

> On our CI, we are testing our changes on both iOS and Android, and we are linting our code.

Reviewer: ```suggestion
On our CI, we test our changes on both iOS and Android, and we lint-check our code.
```
